jerry-core/ecma/base/ecma-globals.h

```diff
@@ -204,20 +204,97 @@ typedef enum
+  ECMA_PROPERTY_TYPE__MAX = ECMA_PROPERTY_TYPE_NAMEDACCESSOR, /**< highest value for property types. */
+} ecma_property_types_t;
+
+/**
+ * Property type mask.
+ */
+#define ECMA_PROPERTY_TYPE_MASK 0x7
+
+/**
+ * Property flags base shift.
+ */
+#define ECMA_PROPERTY_FLAG_SHIFT 3
+
+/**
+ * Property flag list (for ECMA_PROPERTY_TYPE_NAMEDDATA
+ * and ECMA_PROPERTY_TYPE_NAMEDACCESSOR).
  */
 typedef enum
 {
-  ECMA_PROPERTY_FLAG_NAMEDDATA = 1u << 0, /**< property is named data */
-  ECMA_PROPERTY_FLAG_NAMEDACCESSOR = 1u << 1, /**< property is named accessor */
-  ECMA_PROPERTY_FLAG_INTERNAL = 1u << 2, /**< property is internal property */
-  ECMA_PROPERTY_FLAG_CONFIGURABLE = 1u << 3, /**< property is configurable */
-  ECMA_PROPERTY_FLAG_ENUMERABLE = 1u << 4, /**< property is enumerable */
-  ECMA_PROPERTY_FLAG_WRITABLE = 1u << 5, /**< property is writable */
-  ECMA_PROPERTY_FLAG_LCACHED = 1u << 6, /**< property is lcached */
+  ECMA_PROPERTY_FLAG_CONFIGURABLE = 1u << (ECMA_PROPERTY_FLAG_SHIFT + 0), /**< property is configurable */
+  ECMA_PROPERTY_FLAG_ENUMERABLE = 1u << (ECMA_PROPERTY_FLAG_SHIFT + 1), /**< property is enumerable */
+  ECMA_PROPERTY_FLAG_WRITABLE = 1u << (ECMA_PROPERTY_FLAG_SHIFT + 2), /**< property is writable */
+  ECMA_PROPERTY_FLAG_LCACHED = 1u << (ECMA_PROPERTY_FLAG_SHIFT + 3), /**< property is lcached */
 } ecma_property_flags_t;
 
 /**
+ * Abstract property representation.
+ *
+ * A property is a type_and_flags byte and an ecma_value_t value pair.
+ * This pair is represented by a single pointer in JerryScript. Although
+ * a packed struct would only consume sizeof(ecma_value_t)+1 memory
+ * bytes, accessing such structure is inefficient from the CPU viewpoint
+ * because the value is not naturally aligned. To improve performance,
+ * multiple type bytes and values are packed together. The maximum
+ * number of packed items is sizeof(ecma_value_t). The memory layout is
+ * the following when the maximum number of items is present:
+ *
+ *  [type 1, type 2, type 3, type 4][value 1][value 2][value 3][value 4]
+ *
+ * This way no memory is wasted and values are naturally aligned.
+ *
+ * For property pairs, only two values are used:
+ *
+ *  [type 1, type 2, unused 1, unused 2][value 1][value 2]
+ *
+ * The unused two bytes are used to store a compressed pointer for the
+ * next property pair.
+ *
+ * The advantage of this layout is that the value reference can be computed
+ * from the property address. However, property pointers cannot be compressed
+ * anymore.
+ */
+typedef struct
+{
+  uint8_t type_and_flags; /**< ecma_property_types_t (3 bit) and ecma_property_flags_t */
+} ecma_property_t;
+
+/**
+ * Property header for all items in a property list.
+ */
+typedef struct
+{
+  ecma_property_t types[2]; /**< two property type slot. The first represent
+                             *   the type of this property (e.g. property pair) */
+  mem_cpointer_t next_property_cp; /**< next cpointer */
+} ecma_property_header_t;
```

Comment:
It cannot, because ecma_property_header_t must be 4 byte long. I have no plans to change it in the near future. You can work on extending this if you wish.

The point of the header is that everything in the property chain must start with it, and algorithms can ignore those property items which are unknown to them.

```
struct {
  ecma_property_header_t header;
  fields
} hash;
```